File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> mp;
        stack<int> st;
        for(int n:nums2){
            while(!st.empty() && st.top()<n){
                mp[st.top()]=n;
                st.pop();
            }
            st.push(n);
        }

        while(!st.empty()){
            mp[st.top()]=-1;
        st.pop();
        }
        vector<int> res(nums1.size());
        for(int i=0;i<nums1.size();i++){
            res[i]=mp[nums1[i]];
        }
    return res;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer,Integer>mp=new HashMap<>();
        Stack<Integer> st=new Stack<>();
        for(int n:nums2){
            while(!st.isEmpty() && st.peek()<n){
                mp.put(st.pop(),n);
            }
            st.push(n);
        }

        while(!st.isEmpty()){
            mp.put(st.pop(),-1);
        }

        int res[]=new int[nums1.length];
        for(int i=0;i<res.length;i++){
            res[i]=mp.get(nums1[i]);
        }
    return res;
    }
}